File: packages/azure/ios/AzureTranscriptBuilder.swift
import Foundation
import MicrosoftCognitiveServicesSpeech

/// Converts an Azure SDK recognition result into the internal `AzureTranscript`.
/// Parses the `SpeechServiceResponse_JsonResult` property when detailed output
/// is requested, to populate per-word segments + aggregate confidence.
///
/// Detailed JSON shape (when `OutputFormat.detailed` + `requestWordLevelTimestamps`):
/// ```
/// {
///   "RecognitionStatus": "Success",
///   "Offset": 1230000,           // 100-ns ticks
///   "Duration": 5670000,
///   "DisplayText": "Hello world.",
///   "NBest": [{
///     "Confidence": 0.92,
///     "Lexical": "hello world",
///     "Display": "Hello world.",
///     "Words": [
///       {"Word": "hello", "Offset": 1230000, "Duration": 2230000, "Confidence": 0.94},
///       ...
///     ]
///   }]
/// }
/// ```
internal enum AzureTranscriptBuilder {

  /// Build a transcript from an Azure recognition event result. Accepts the
  /// base `SPXRecognitionResult` so both `SPXSpeechRecognitionResult` and
  /// `SPXConversationTranscriptionResult` flow through the same code path —
  /// the diarization-aware speakerId is supplied by the caller.
  static func from(
    result: SPXRecognitionResult,
    isFinal: Bool,
    sessionId: String,
    sessionStartMs: Double,
    fallbackLanguage: String,
    speakerId: String = ""
  ) -> AzureTranscript {
    let nowMs = Date().timeIntervalSince1970 * 1_000
    let text = result.text ?? ""

    // Azure offset/duration are in 100-ns ticks → convert to ms (÷ 10,000).
    let resultOffsetMs = Double(result.offset) / 10_000.0
    let resultDurationMs = Double(result.duration) / 10_000.0

    // Parse the detailed JSON payload if present.
    let jsonRaw = result.properties?.getPropertyBy(SPXPropertyId.speechServiceResponseJsonResult) ?? ""
    let parsed = parseDetailedJson(jsonRaw)

    let segments = parsed?.words ?? []
    let confidence = parsed?.confidence ?? -1
    let detectedLanguage = parsed?.language ?? fallbackLanguage

    return AzureTranscript(
      id: UUID().uuidString,
      sessionId: sessionId,
      timestamp: nowMs,
      text: text,
      isFinal: isFinal,
      language: detectedLanguage,
      confidence: confidence,
      offsetMs: resultOffsetMs >= 0 ? resultOffsetMs : (nowMs - sessionStartMs),
      durationMs: resultDurationMs > 0 ? resultDurationMs : -1,
      speakerId: speakerId,
      segments: segments
    )
  }

  /// Build a transcript when we only have a text string (stitched session
  /// transcript on stop(), or discard with last partial).
  static func fromText(
    _ text: String,
    isFinal: Bool,
    sessionId: String,
    sessionStartMs: Double,
    language: String
  ) -> AzureTranscript {
    let nowMs = Date().timeIntervalSince1970 * 1_000
    return AzureTranscript(
      id: UUID().uuidString,
      sessionId: sessionId,
      timestamp: nowMs,
      text: text,
      isFinal: isFinal,
      language: language,
      confidence: -1,
      offsetMs: nowMs - sessionStartMs,
      durationMs: -1,
      speakerId: "",
      segments: []
    )
  }

  // MARK: JSON parsing

  private struct ParsedDetailed {
    let words: [AzureSegment]
    let confidence: Double
    let language: String?
  }

  private static func parseDetailedJson(_ raw: String) -> ParsedDetailed? {
    guard !raw.isEmpty, let data = raw.data(using: .utf8) else { return nil }
    guard let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
      return nil
    }

    // Some payloads expose "PrimaryLanguage" when auto-detect is on.
    let language: String? = (obj["PrimaryLanguage"] as? [String: Any])?["Language"] as? String

    guard let nBest = obj["NBest"] as? [[String: Any]], let first = nBest.first else {
      return nil
    }

    let confidence: Double = (first["Confidence"] as? Double) ?? -1
    let wordsJson = first["Words"] as? [[String: Any]] ?? []

    let words: [AzureSegment] = wordsJson.compactMap { w in
      guard let text = w["Word"] as? String else { return nil }
      let offsetTicks = (w["Offset"] as? Double) ?? 0
      let durationTicks = (w["Duration"] as? Double) ?? 0
      let wConfidence = (w["Confidence"] as? Double) ?? -1
      return AzureSegment(
        text: text,
        startMs: offsetTicks / 10_000.0,
        durationMs: durationTicks / 10_000.0,
        confidence: wConfidence,
        alternatives: []
      )
    }

    return ParsedDetailed(words: words, confidence: confidence, language: language)
  }
}

File: packages/azure/android/src/main/java/com/clarionhq/azure/AzureTranscriptBuilder.kt
package com.clarionhq.azure

import com.clarionhq.azure.AzureConstants.TICKS_PER_MS
import com.clarionhq.azure.AzureConstants.UNKNOWN
import com.microsoft.cognitiveservices.speech.PropertyId
import com.microsoft.cognitiveservices.speech.RecognitionResult
import org.json.JSONException
import org.json.JSONObject
import java.util.UUID

/**
 * Converts an Azure SDK recognition result into [AzureTranscript].
 * Parses the `SpeechServiceResponse_JsonResult` property when detailed output
 * is requested, to populate per-word segments + aggregate confidence.
 *
 * The detailed JSON payload exposes `NBest[0].Words` (per-word offset/duration/confidence
 * in ticks — see [AzureConstants.TICKS_PER_MS] for the ticks→ms conversion) and
 * `PrimaryLanguage.Language` when auto-detect is enabled.
 */
internal object AzureTranscriptBuilder {

  /**
   * Build a transcript from an Azure recognition result. Accepts the base
   * [RecognitionResult] so both `SpeechRecognitionResult` and
   * `ConversationTranscriptionResult` flow through the same code path — the
   * diarization-aware speakerId is supplied by the caller.
   */
  fun from(
    result: RecognitionResult,
    isFinal: Boolean,
    sessionId: String,
    sessionStartMs: Double,
    fallbackLanguage: String,
    speakerId: String = "",
  ): AzureTranscript {
    val text = result.text ?: ""
    val base = fromText(
      text = text,
      isFinal = isFinal,
      sessionId = sessionId,
      sessionStartMs = sessionStartMs,
      language = fallbackLanguage,
    )

    val jsonRaw = result.properties?.getProperty(PropertyId.SpeechServiceResponse_JsonResult) ?: ""
    val parsed = parseDetailedJson(jsonRaw)

    val rawOffsetMs = ticksToMs(result.offset?.toLong() ?: 0L)
    val rawDurationMs = ticksToMs(result.duration?.toLong() ?: 0L)
    val offsetMs = if (rawOffsetMs > 0) rawOffsetMs else base.offsetMs
    val durationMs = if (rawDurationMs > 0) rawDurationMs else UNKNOWN

    return base.copy(
      language = parsed?.language ?: fallbackLanguage,
      confidence = parsed?.confidence ?: UNKNOWN,
      offsetMs = offsetMs,
      durationMs = durationMs,
      speakerId = speakerId,
      segments = parsed?.words ?: emptyList(),
    )
  }

  /** Build a transcript from a plain text string (stitched session-final, discard fallback, etc.). */
  fun fromText(
    text: String,
    isFinal: Boolean,
    sessionId: String,
    sessionStartMs: Double,
    language: String,
  ): AzureTranscript {
    val nowMs = System.currentTimeMillis().toDouble()
    return AzureTranscript(
      id = UUID.randomUUID().toString(),
      sessionId = sessionId,
      timestamp = nowMs,
      text = text,
      isFinal = isFinal,
      language = language,
      confidence = UNKNOWN,
      offsetMs = nowMs - sessionStartMs,
      durationMs = UNKNOWN,
      speakerId = "",
      segments = emptyList(),
    )
  }

  // MARK: helpers

  private fun ticksToMs(ticks: Long): Double = ticks / TICKS_PER_MS

  // MARK: JSON parsing

  private data class ParsedDetailed(
    val words: List<AzureSegment>,
    val confidence: Double,
    val language: String?,
  )

  private fun parseDetailedJson(raw: String): ParsedDetailed? {
    if (raw.isEmpty()) return null
    return try {
      val obj = JSONObject(raw)

      val language: String? = obj.optJSONObject("PrimaryLanguage")
        ?.optString("Language")?.takeIf { it.isNotEmpty() }

      val nBest = obj.optJSONArray("NBest") ?: return null
      if (nBest.length() == 0) return null
      val first = nBest.optJSONObject(0) ?: return null

      val confidence = first.optDouble("Confidence", UNKNOWN)
      val wordsArr = first.optJSONArray("Words")
      val words = mutableListOf<AzureSegment>()
      if (wordsArr != null) {
        for (i in 0 until wordsArr.length()) {
          val w = wordsArr.optJSONObject(i) ?: continue
          val wText = w.optString("Word", "")
          if (wText.isEmpty()) continue
          words += AzureSegment(
            text = wText,
            startMs = ticksToMs(w.optLong("Offset", 0L)),
            durationMs = ticksToMs(w.optLong("Duration", 0L)),
            confidence = w.optDouble("Confidence", UNKNOWN),
            alternatives = emptyList(),
          )
        }
      }
      ParsedDetailed(words = words, confidence = confidence, language = language)
    } catch (_: JSONException) {
      null
    }
  }
}
